### PR TITLE
OCPBUGS-31663: UPSTREAM: <carry>: add management workload check for guaranteed qos

### PR DESCRIPTION
**Pull request by eggfoobar**
when static pods have workload partitioning enabled we should not alter their resources if they are Guaranteed QoS, this change adds a check for Guaranteed QoS

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by eggfoobar** [comment]
Validates Locally:

Created a cluster with workload partitioning. Connected to the host via a debug pod and created a static pod definition in `/etc/kuberentes/manifests/`

The result showed the correct warning annotation being applied and the QoS remaining as `Guaranteed`

Static Pod Manifest:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: busybox
  namespace: openshift-machine-config-operator
  annotations:
    target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
spec:
  containers:
  - args:
    - sleep
    - "3600"
    image: busybox
    imagePullPolicy: IfNotPresent
    name: limitbox
    resources:
      limits:
        cpu: 20m
        memory: 50Mi
      requests:
        cpu: 20m
        memory: 50Mi
  ```

Running Pod:

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
   ...
    workload.openshift.io/warning: skipping pod CPUs requests modifications because
      it has guaranteed QoS class
  name: busybox-ip-10-0-1-71.us-west-2.compute.internal
  namespace: openshift-machine-config-operator
...
spec:
  containers:
  - args:
    - sleep
    - "3600"
    image: busybox
    imagePullPolicy: IfNotPresent
    name: limitbox
    resources:
      limits:
        cpu: 20m
        memory: 50Mi
      requests:
        cpu: 20m
        memory: 50Mi
 ...
status:
  ...
  qosClass: Guaranteed
```

File: pkg/kubelet/managed/managed.go [hunk 1]
```diff
@@ -139,3 +149,50 @@ func updateContainers(workloadName string, pod *v1.Pod) error {
...
+// isStaticPodGuaranteed checks if the static pod has a guaranteed QoS.
+// This QoS check is different from the library versions that check QoS,
+// this is because of the order at which changes get observed.
+// (i.e. the library assumes the defaulter has ran on the pod resource before calculating QoS).
+//
+// The static files will get interpreted before they reach the API server and before the defaulter applies changes,
+// this function takes into account the case where only `limits.cpu` are provided but no `requests.cpu` are since that
+// counts as a Guaranteed QoS after the defaulter runs.
+func isStaticPodGuaranteed(pod *v1.Pod) bool {
+	isGuaranteed := func(containers []v1.Container) bool {
```
[the remark this answers is not in the document]
Comment (eggfoobar) [reply pkg/kubelet/managed/managed.go:172]: Oh, that's a great catch! It did make me rethink this function a bit, there is no uniqueness to static pods in this check, so I removed the `static` keyword from it. I took your advice and added that check to the public function that kicks things off here https://github.com/openshift/kubernetes/pull/1928/files#diff-e5bdbb2def4921c3397e7d994bad5224ed876b85d5d1c7656a808f5f6da7f848R92-R95

**Comment by eggfoobar** [comment]
We needed to remove the static pod check that we added. 

After some testing, the way that `kubelet` reads in static pod files means that at the time of the call site to [ModifyStaticPodForPinnedManagement](https://github.com/openshift/kubernetes/blob/8628c3c4da6746b1dc967cc520b189a04ebd78d1/pkg/kubelet/config/file.go#L235) we do not know if the pod is static or not since Kubelet will read in all the files first. Then they are placed in the [pod config cache](https://github.com/openshift/kubernetes/blob/5a51bd8afbb871ff9952c94788d368d6254cd500/pkg/kubelet/config/config.go#L76) via a [channel](https://github.com/openshift/kubernetes/blob/5a51bd8afbb871ff9952c94788d368d6254cd500/pkg/kubelet/config/file.go#L79) that triggers the [annotation code](https://github.com/openshift/kubernetes/blob/5a51bd8afbb871ff9952c94788d368d6254cd500/pkg/kubelet/config/config.go#L245) to be applied. This means we can't actually check if a pod is static at this stage with out more invasive changes that should be done in a different PR if desired.

**Comment by eggfoobar** [comment]
/retest-required

Unit test failures do not seem related to this change.

File: pkg/kubelet/managed/managed.go [hunk 2]
```diff
@@ -145,3 +155,50 @@ func updateContainers(workloadName string, pod *v1.Pod) error {
...
+			for _, r := range []v1.ResourceName{v1.ResourceMemory, v1.ResourceCPU} {
+				limit := c.Resources.Limits[r]
+				request, requestExist := c.Resources.Requests[r]
+				if limit.IsZero() {
+					return false
+				}
+				if !requestExist {
+					continue
+				}
+				// it some corner case, when you set CPU request to 0 the k8s will change it to the value
```
Comment (mrunalp) [inline pkg/kubelet/managed/managed.go:180]: nit: it -> In, the -> then

File: pkg/kubelet/managed/managed_test.go [hunk 3]
```diff
@@ -247,6 +329,86 @@ func TestStaticPodThrottle(t *testing.T) {
...
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodPending,
+				},
+			},
+			expectedAnnotations: map[string]string{
+				WorkloadAnnotationWarning: qosWarning,
+			},
+			isGuaranteed: true,
+		},
```
Comment (mrunalp) [inline pkg/kubelet/managed/managed_test.go:411]: Should we add any negative test cases?

File: pkg/kubelet/managed/managed.go (hunk 2 above)
Comment (eggfoobar) [reply pkg/kubelet/managed/managed.go]: Thanks, fixed!

File: pkg/kubelet/managed/managed_test.go (hunk 3 above)
Comment (eggfoobar) [reply pkg/kubelet/managed/managed_test.go:612]: Good point, I added some more unit tests to cover those, and added an error state unit tests as well, now we have a `94.7%` coverage on the `managed.go` file

**Pull request merged**